python/eight_mile/pytorch/layers.py

```diff
@@ -1434,3 +1516,227 @@ def call(self, inputs):
+
+        :param encoder_outputs: `namedtuple` The outputs of the encoder class.
+        :param init: `Callable(ecnoder_outputs: encoder_outputs, K: int)` -> Any: A
+            callable that is called once at the start of the search to initialize
+            things. This returns a blob that is passed to other callables.
+        :param step: `Callable(paths: torch.LongTensor, extra) -> (probs: torch.FloatTensor, extra):
+            A callable that is does a single decoding step. It returns the log
+            probabilities over the vocabulary in the last dimension. It also returns
+            any state the decoding process needs.
+        :param update: `Callable(beams: torch.LongTensor, extra) -> extra:
+            A callable that is called to edit the decoding state based on the selected
+            best beams.
+        :param length_penalty: `Callable(lengths: torch.LongTensor) -> torch.floatTensor
+            A callable that generates a penalty based on the lengths. Lengths is
+            [B, K] and the returned penalty should be [B, K, 1] (or [B, K, V] to
+            have token based penalties?)
+
+        :Keyword Arguments:
+        * *beam* -- `int`: The number of beams to use.
+        * *mxlen* -- `int`: The max number of steps to run the search for.
+
+        :returns:
+            tuple(preds: torch.LongTensor, lengths: torch.LongTensor, scores: torch.FloatTensor)
+            preds: The predicted values: [B, K, max(lengths)]
+            lengths: The length of each prediction [B, K]
+            scores: The score of each path [B, K]
+        """
+        mxlen = kwargs.get('mxlen', 100)
+        bsz = encoder_outputs.output.shape[0]
+        device = encoder_outputs.output.device
+        with torch.no_grad():
+            extra = self.init(encoder_outputs)
+            paths = torch.full((bsz, self.K, 1), Offsets.GO, dtype=torch.long, device=device)
+            # This tracks the log prob of each beam. This is distinct from score which
+            # is based on the log prob and penalties.
+            log_probs = torch.zeros((bsz, self.K), dtype=torch.float, device=device)
+            # Tracks the lengths of the beams, unfinished beams have lengths of zero.
+            lengths = torch.zeros((bsz, self.K), dtype=torch.long, device=device)
+
+            for i in range(mxlen - 1):
+                probs, extra = self.step(paths, extra)
+                V = probs.shape[-1]
+                probs = probs.view((bsz, self.K, V))  # [B, K, V]
+                if i > 0:
+                    # This mask is for all beams that are done.
+                    done_mask = (lengths != 0).unsqueeze(-1)  # [B, K, 1]
+                    # Can creating this mask be moved out of the loop? It never changes but we don't have V
+                    # This mask selects the EOS token
+                    eos_mask = torch.zeros((1, 1, V), dtype=MASK_TYPE, device=device)
```

Comment:
I thought the plan was to remove the `MASK_TYPE` and have the dtype based on the dtype for `done_mask` which should correctly handle the type for different versions of pytorch? Or did that not work?

Reply:
yep. its on the list!  moving slower than I hoped